src/thread_id.rs

```diff
@@ -97,17 +106,41 @@ cfg_if::cfg_if! {
                 // will go through get_slow which will either panic or
                 // initialize a new ThreadGuard.
                 unsafe {
-                    THREAD = None;
+                    (&mut *THREAD.get()).self_ptr = null();
+                }
+                THREAD_ID_MANAGER.lock().free(self.id.get());
+            }
+        }
+
+        /// Data which is unique to the current thread while it is running.
+        /// A thread ID may be reused after a thread exits.
+        ///
+        /// This wrapper exists to hide multiple accesses to the TLS data
+        /// from the backend as this can lead to inefficient codegen
+        /// (to be precise it can lead to multiple TLS address lookups)
+        #[derive(Clone, Copy)]
+        struct ThreadWrapper {
+            self_ptr: *const Thread,
```

Comment:
What is the purpose of `self_ptr`? It seems to always point to the `thread` field. Is this just a `bool` that indicates whether the thread has been initialized? How is this better than just using an `Option`?

Reply:
No, that is what i was describing earlier, this is the hack that tricks the compiler into actually only ever generating one `get_tls_addr` call and reusing the calculated address, if the compiler was smart enough, this `self_ptr` wouldn't be necessary at all.